### PR TITLE
Set difference when picking random ports

### PR DESCRIPTION
**Pull request by dadgar**
The way that dynamic port assignment is done currently doesn't work well at all if significant ranges of reserved_ports are specified in nomad client config. For example:

```
client {
  reserved {
    reserved_ports = "23000-65535" 
  }
}
```
With this config (which leaves ports 20000-22999 available for assignment), running a job requiring several dynamic ports will fail more often than not, as a random pick in the range 20000-60000 will fall in a reserved range >90% of the time.

This change ensures that acceptable ports are always assigned, so long as a sufficient number of free ports exist in any non-reserved range between MinDynamicPort (20000) and MaxDynamicPort (60000).

Thanks to @bagelswitch for reporting.

@armon  for review

### CODE REVIEW

File: nomad/structs/bitmap.go [hunk 1]
```diff
@@ -17,6 +17,32 @@ func NewBitmap(size int) (Bitmap, error) {
...
+		return nil, fmt.Errorf("can't copy nil Bitmap")
+	}
+
+	nb, err := NewBitmap(b.Size())
+	if err != nil {
+		return nil, err
+	}
+
+	s := b.Size()
+	for i := uint(0); i < s; i++ {
```
Comment (armon) [inline nomad/structs/bitmap.go]: I would copy full bytes at a time not bits. You have access to the underlying byte array, and there is no need to add a conditional set.

File: nomad/structs/network.go [hunk 2]
```diff
@@ -208,28 +209,25 @@ func (idx *NetworkIndex) AssignNetwork(ask *NetworkResource) (out *NetworkResour
...
-				err = fmt.Errorf("dynamic port selection failed")
-				return
-			}
+		// Try to stochastically pick the dynamic ports as it is faster and
+		// lower memory usage.
+		var dynPorts []int
+		var dynErr error
+		dynPorts, dynErr = getDynamicPortsStochastic(used, ask)
+		if err != nil {
+			goto BUILD_OFFER
```
Comment (armon) [inline nomad/structs/network.go]: If you get `fmt.Errorf("stochastic dynamic port selection failed")` you are jumping over `getDynamicPortsPrecise`

File: nomad/structs/network.go [hunk 3]
```diff
@@ -208,28 +209,25 @@ func (idx *NetworkIndex) AssignNetwork(ask *NetworkResource) (out *NetworkResour
...
+		}
 
-			randPort := MinDynamicPort + rand.Intn(MaxDynamicPort-MinDynamicPort)
-			used := idx.UsedPorts[ipStr]
-			if used != nil && used.Check(uint(randPort)) {
-				goto PICK
-			}
+		// Fall back to the precise method if the random sampling failed.
+		dynPorts, dynErr = getDynamicPortsPrecise(used, ask)
```
Comment (armon) [inline nomad/structs/network.go]: Doesn't this overwrite the `dynPorts`?

**Comment by armon** [comment]
left some comments, LGTM

File: nomad/structs/network.go (hunk 3 above)
Comment (dadgar) [inline nomad/structs/network.go]: Yeah I want it to. We only get here if the stochastic method failed

**Pull request merged**